### PR TITLE
Customize runner type

### PR DESCRIPTION
**Pull request by yhcao6**
Add support to customize runner type ('EpochBasedRunner', 'IterBasedRunner') and merge the 'total_epoch' attribute in the config with the newly added runner attribute as follows:
```
runner = dict(type='EpochBasedRunner', max_epochs=12)
```

### CODE REVIEW

File: configs/faster_rcnn/faster_rcnn_r50_caffe_fpn_mstrain_90000it_coco.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    warmup_iters=500,
+    warmup_ratio=0.001,
+    step=[60000, 80000])
+total_epochs = 90000
+
+# Runner type
+runner = 'IterBasedRunner'
+
+checkpoint_config = dict(interval=10000)
+evaluation = dict(interval=10000, metric='bbox')
```
Comment (ZwwWayne) [inline configs/faster_rcnn/faster_rcnn_r50_caffe_fpn_mstrain_90000it_coco.py:16]: The configuration of runner should be a dict, see examples in [mmcls here](https://github.com/open-mmlab/mmclassification/blob/master/configs/_base_/schedules/imagenet_bs2048.py#L12).

File: configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py [hunk 2]
```diff
@@ -32,7 +32,7 @@
     warmup_ratio=0.001,
     # [7] yields higher performance than [6]
     step=[7])
-total_epochs = 8  # actual epoch = 8 * 8 = 64
+runner = dict(max_epochs=8)
```
Comment (ZwwWayne) [inline configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py:35]: comments should be kept.

File: configs/cityscapes/mask_rcnn_r50_fpn_1x_cityscapes.py [hunk 3]
```diff
@@ -39,7 +39,7 @@
     warmup_ratio=0.001,
     # [7] yields higher performance than [6]
     step=[7])
-total_epochs = 8  # actual epoch = 8 * 8 = 64
```
Comment (ZwwWayne) [inline configs/cityscapes/mask_rcnn_r50_fpn_1x_cityscapes.py:42]: comments should be kept.

File: configs/faster_rcnn/faster_rcnn_r50_caffe_fpn_mstrain_90000it_coco.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+_base_ = 'faster_rcnn_r50_caffe_fpn_mstrain_1x_coco.py'
```
Comment (ZwwWayne) [inline configs/faster_rcnn/faster_rcnn_r50_caffe_fpn_mstrain_90000it_coco.py:1]: May directly use 90k as those in mmseg.

File: configs/pascal_voc/faster_rcnn_r50_fpn_1x_voc0712.py [hunk 5]
```diff
@@ -10,4 +10,4 @@
 # actual epoch = 3 * 3 = 9
 lr_config = dict(policy='step', step=[3])
 # runtime settings
-total_epochs = 4  # actual epoch = 4 * 3 = 12
```
Comment (ZwwWayne) [inline configs/pascal_voc/faster_rcnn_r50_fpn_1x_voc0712.py:13]: keep comments.

**Comment by ZwwWayne** [comment]
LGTM after resolving comments. Please @jshilong also have a look.

File: mmdet/apis/train.py [hunk 6]
```diff
@@ -84,12 +86,25 @@ def train_detector(model,
...
     # build runner
     optimizer = build_optimizer(model, cfg.optimizer)
-    runner = EpochBasedRunner(
-        model,
-        optimizer=optimizer,
-        work_dir=cfg.work_dir,
-        logger=logger,
-        meta=meta)
+
+    if cfg.get('runner') is None:
```
Comment (jshilong) [inline mmdet/apis/train.py:90]: `if "runner" in cfg:` may be better.

File: mmdet/apis/train.py [hunk 7]
```diff
@@ -84,12 +86,25 @@ def train_detector(model,
...
+        cfg.runner = {
+            'type': 'EpochBasedRunner',
+            'max_epochs': cfg.total_epochs
+        }
+        warnings.warn(
```
Comment (jshilong) [inline mmdet/apis/train.py:95]: During the interface change period, there may be a situation where `total_epochs` and `max_epochs` in the runner exist at the same time, we would better assert this case.

File: mmdet/core/evaluation/eval_hooks.py [hunk 8]
```diff
@@ -138,26 +140,37 @@ def evaluation_flag(self, runner):
...
+            self.save_best(runner, key_score)
+
+    def after_train_iter(self, runner):
+        if self.by_epoch or not self.every_n_iters(runner, self.interval):
+            return
+        from mmdet.apis import single_gpu_test
+        results = single_gpu_test(runner.model, self.dataloader, show=False)
+        key_score = self.evaluate(runner, results)
+        if self.save_best:
+            self.save_best(runner, key_score)
```
Comment (jshilong) [inline mmdet/core/evaluation/eval_hooks.py:158]: `save_best_cpt` ?

File: mmdet/core/evaluation/eval_hooks.py [hunk 9]
```diff
@@ -138,26 +140,37 @@ def evaluation_flag(self, runner):
...
+
+    def save_best_cpt(self, runner, key_score):
```
Comment (jshilong) [inline mmdet/core/evaluation/eval_hooks.py:160]: `cpt` may be confusing, `save_best` or `save_best_checkpoint` may be better.

File: mmdet/core/evaluation/eval_hooks.py [hunk 10]
```diff
@@ -243,9 +258,41 @@ def after_train_epoch(self, runner):
...
+        if self.by_epoch or not self.every_n_iters(runner, self.interval):
             return
 
+        # Synchronization of BatchNorm's buffer (running_mean
+        # and running_var) is not supported in the DDP of pytorch,
+        # which may cause the inconsistent performance of models in
+        # different ranks, so we broadcast BatchNorm's buffers
+        # of rank 0 to other ranks to avoid this.
+        if self.broadcast_bn_buffer:
+            model = runner.model
```
Comment (jshilong) [inline mmdet/core/evaluation/eval_hooks.py:289]: Seems this code snippet is used several times, how about turning it into a method?

File: mmdet/core/evaluation/eval_hooks.py [hunk 11]
```diff
@@ -243,7 +258,11 @@ def after_train_epoch(self, runner):
                     dist.broadcast(module.running_var, 0)
                     dist.broadcast(module.running_mean, 0)
 
-        if not self.evaluation_flag(runner):
+    def after_train_epoch(self, runner):
+        if self.broadcast_bn_buffer:
```
Comment (jshilong) [inline mmdet/core/evaluation/eval_hooks.py:262]: We can move `broadcast_bn_buffer` after 
```
 if not self.by_epoch or not self.evaluation_flag(runner):
            return	      
```` 
like `after_train_iter`

**Review by jshilong — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

File: mmdet/apis/train.py [hunk 12]
```diff
@@ -84,12 +86,28 @@ def train_detector(model,
...
+
+    if 'runner' not in cfg:
+        cfg.runner = {
+            'type': 'EpochBasedRunner',
+            'max_epochs': cfg.total_epochs
+        }
+        warnings.warn(
+            'config is now expected to have a `runner` section, '
+            'please set `runner` in your config.', UserWarning)
+    else:
```
Comment (ZwwWayne) [inline mmdet/apis/train.py:98]: Can we use `elif` here?

File: docs/tutorials/config.md [hunk 13]
```diff
@@ -395,6 +395,7 @@ lr_config = dict(  # Learning rate scheduler config used to register LrUpdater h
     0.001,  # The ratio of the starting learning rate used for warmup
     step=[8, 11])  # Steps to decay the learning rate
 total_epochs = 12  # Total epochs to train the model
```
Comment (ZwwWayne) [inline docs/tutorials/config.md:397]: This line should be removed or should be indicated as unnecessary in the comments.

**Comment by ZwwWayne** [comment]
May also enrich the PR message to describe the supported feature and migration code in detail. Thus other users could understand the PR more clearly as this modification is important.

**Pull request merged**